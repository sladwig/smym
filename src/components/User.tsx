import React, { useState, useEffect } from 'react';
import { IUser } from '../store/User';
import './user.css';
import { TransactionList } from './TranactionList';
import { Icon } from './Icon';
import { PaidButton, UnpaidButton } from './Button';
import classnames from 'classnames';

interface IProps {
    user: IUser;
    expand: boolean;
}
const transactionHeight = 102;

export const User = ({ user, expand }: IProps) => {
    const [showMore, setShowMore] = useState(expand);
    useEffect(() => setShowMore(expand), [expand]);

    const height = showMore ? (user.transactions.length || 1) * transactionHeight + 91 : 56;

    const toggleShowMore = () => setShowMore(() => !showMore);
    return (
        <div className="user-wrapper">
            <div className="user-area" style={{ height }}>
                <div className="user">
                    <Avatar onClick={toggleShowMore} src={user.avatar} />
                    <div className="description" onClick={toggleShowMore}>
                        <b style={{ marginRight: 12 }}>{user.name}</b> |{' '}
                        <span style={{ marginLeft: 12 }}>{user.real_name}</span>
                    </div>
                    <div className="balance" style={{ marginRight: 26 }}>
                        {user.balance.toFixed(2)}
                    </div>
                    {user.balance === 0 ? <PaidButton /> : <UnpaidButton />}
                    <Icon
                        name="open"
                        transparent={true}
                        turn={showMore}
                        noHover={true}
                        pointer={true}
                        style={{ marginLeft: 15 }}
                        onClick={toggleShowMore}
                    />
                </div>
                {showMore && <TransactionList transactions={user.transactions} />}
            </div>
        </div>
    );
};
interface AvatarProps {
    src?: string;
    onClick?: () => any;
    smaller?: boolean;
}
export const Avatar = ({ src, onClick, smaller = false }: AvatarProps) => (
    <div className={classnames('avatar-area', { smaller })} onClick={onClick}>
        <img className="avatar" src={src} />
    </div>
);
