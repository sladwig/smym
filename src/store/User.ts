import { types, Instance } from 'mobx-state-tree';
import { Transaction } from './Transaction';

export const User = types
    .model('User', {
        id: types.identifier,
        name: types.maybe(types.string),
        real_name: types.maybe(types.string),
        dmChannel: types.maybe(types.string),
        transactions: types.array(Transaction),
    })
    .actions(self => ({
        setDmChannel(channelId: string) {
            self.dmChannel = channelId;
        },
        addTransaction(value: number | 'reset', description: string) {
            if (value === 'reset') return this.resetTransactions();
            self.transactions.push({ value, description });
        },
        resetTransactions() {
            self.transactions.clear();
        },
        update(user: any) {
            self.name = user.name;
            self.real_name = user.real_name;
        },
    }))
    .views(self => ({
        get balance() {
            return self.transactions.reduce((sum, current) => sum + current.value, 0);
        },
    }));

export type IUser = Instance<typeof User>;
