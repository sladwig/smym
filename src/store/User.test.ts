import { User } from './User';

test('can be created', () => {
    const user = User.create({ id: 'abc' });
    expect(user).toMatchSnapshot();
});
