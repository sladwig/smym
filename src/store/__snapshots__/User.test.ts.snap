// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can be created 1`] = `
Object {
  "avatar": undefined,
  "dmChannel": undefined,
  "id": "abc",
  "name": undefined,
  "real_name": undefined,
  "transactions": Array [],
}
`;
